homeassistant/components/nextcloud/sensor.py

```diff
@@ -0,0 +1,111 @@
+import logging
+
+from nextcloudmonitor import NextcloudMonitor
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_URL,
+    CONF_USERNAME,
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import track_time_interval
+
+_LOGGER = logging.getLogger(__name__)
+
+# Assign domain
+DOMAIN = "nextcloud"
+# Set Default scan interval in seconds
+SCAN_INTERVAL = timedelta(seconds=60)
+
+# Validate user configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_URL): cv.url,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL): cv.time_period,
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None) -> None:
+    """Set up the Nextcloud sensor."""
+    # Fetch Nextcloud Monitor api data
+    ncm = NextcloudMonitor(
+        config[CONF_URL], config[CONF_USERNAME], config[CONF_PASSWORD]
+    )
+    hass.data[DOMAIN] = get_sensors(ncm.data)
+
+    def nextcloud_update(event_time):
+        """Update data from nextcloud api."""
+        ncm.update()
+        hass.data[DOMAIN] = get_sensors(ncm.data)
+
+    # Update sensors on time interval
+    track_time_interval(hass, nextcloud_update, config[CONF_SCAN_INTERVAL])
+
+    # Setup sensors
+    sensors = []
+    for name in hass.data[DOMAIN]:
+        sensors.append(NextcloudSensor(name))
+    add_entities(sensors, True)
+
+
+# Use recursion to create list of sensors & values based on nextcloud api data
+def get_sensors(api_data, key_path=""):
+    """Use Recursion to discover sensors and values.
+
+    Get dictionary of sensors by recursing through dict returned by api until
+    the dictionary value does not contain another dictionary and use the
+    resulting path of dictionary keys and resulting value as the name/value
+    for the sensor.
+
+    returns: dictionary of sensors/values
+    """
+    result = {}
+    for key, value in api_data.items():
+        if isinstance(value, dict):
+            key_path += f"{key}_"
+            result.update(get_sensors(value, key_path))
+        else:
+            result[key_path + key] = value
+    return result
+
+
+class NextcloudSensor(Entity):
+    """Represents a Nextcloud sensor."""
+
+    def __init__(self, item):
+        """Initialize the Nextcloud sensor."""
+        self._name = item
+        self._state = None
+
+    @property
+    def icon(self):
+        """Return the icon for this sensor."""
+        return "mdi:cloud"
+
+    @property
+    def name(self):
+        """Return the name for this sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state for this sensor."""
+        return self._state
```

Comment:
What kind of values are we returning as state here?

Reply:
Hi @MartinHjelmare Thanks for reviewing my humble code. The values here can be just about anything that the api returns. Generally floats and strings but I've seen lists and even a dictionary (i think).

Reply:
Different types of state values belong in different kind of platforms in Home Assistant. It would be good to see an overview of the different values the api will return and what they mean, so we can see that we are implementing the integration correctly.

Reply:
Here is a sample api response:

`{"nextcloud":{"system":{"version":"17.0.3.1","theme":"","enable_avatars":"yes","enable_previews":"yes","memcache.local":"\\OC\\Memcache\\Redis","memcache.distributed":"none","filelocking.enabled":"yes","memcache.locking":"\\OC\\Memcache\\Redis","debug":"no","freespace":193857982464,"cpuload":[0.12,0.050000000000000003,0.089999999999999997],"mem_total":4039332,"mem_free":3268096,"swap_total":2097148,"swap_free":2097148,"apps":{"num_installed":44,"num_updates_available":3,"app_updates":{"calendar":"2.0.2","contacts":"3.2.0","tasks":"0.12.1"}}},"storage":{"num_users":2,"num_files":18000,"num_storages":3,"num_storages_local":1,"num_storages_home":2,"num_storages_other":0},"shares":{"num_shares":1,"num_shares_user":0,"num_shares_groups":0,"num_shares_link":1,"num_shares_mail":0,"num_shares_room":0,"num_shares_link_no_password":1,"num_fed_shares_sent":0,"num_fed_shares_received":0,"permissions_3_1":"1"}},"server":{"webserver":"Apache","php":{"version":"7.3.15","memory_limit":536870912,"max_execution_time":3600,"upload_max_filesize":17179869184},"database":{"type":"mysql","version":"5.7.29","size":16285696}},"activeUsers":{"last5minutes":1,"last1hour":2,"last24hours":2}}`

Reply:
Please write it like a table so that it's clear what is what.

At a glance some of these values seem binary in nature, ie they have value like `yes` or `no`. Binary sensors should be part of a binary_sensor platform.

Reply:
How would you suggest going about this? Should we look at the initial value that is returned and use the value to determine the platform type?

Reply:
We need to learn what kind of values the api is giving us and what they mean. We can't assume all of them fit as sensor values in home assistant.

Is there no api documentation?

Reply:
The api project is here:
[https://github.com/nextcloud/serverinfo](https://github.com/nextcloud/serverinfo)

I don't really see anything there that gives any guidelines on what is returned for each item.

Reply:
@meichthys I think you have to look into the actual code like this https://github.com/nextcloud/serverinfo/blob/master/lib/SystemStatistics.php#L71-L86

Reply:
@escoand Hey, thanks!
When i get some time I'll see if i can include BinarySensors where they make sense.

Reply:
I'll be away this weekend but may have time on Monday. I guess I'll have to shoot for v0.108

Reply:
@MartinHjelmare, What do you recommend that we do to keep track of what kind of sensor each piece of data should be? Should we add a Global Variable to track each sensor and its type?

Reply:
Below is a list of the available sensors in the order that they are returned by the api. It appears that there are 4 Binary sensors and the rest seem to be Generic Sensors. It would be easy enough to throw these sensor names into a global variable but it seems like that might make the component a little brittle if the upstream api changes. 

| Sensor Name | Sensor Type |
|---------------|-------------|
|nextcloud_system_version|Generic Sensor|
|nextcloud_system_theme|Generic Sensor|
|nextcloud_system_enable_avitars|Binary Sensor|
|nextcloud_system_enable_previews|Binary Sensor|
|nextcloud_system_memcache.local|Generic Sensor|
|nextcloud_system_memcache.distributed|Generic Sensor|
|nextcloud_system_filelocking.enabled|Binary Sensor|
|nextcloud_system_memcache.locking|Generic Sensor|
|nextcloud_system_debug|Binary Sensor|
|nextcloud_system_freespace|Generic Sensor|
|nextcloud_system_cpuload|Generic Sensor|
|nextcloud_system_mem_total|Generic Sensor|
|nextcloud_system_mem_free|Generic Sensor|
|nextcloud_system_swap_total|Generic Sensor|
|nextcloud_system_swap_free|Generic Sensor|
|nextcloud_system_apps_num_installed|Generic Sensor|
|nextcloud_system_apps_num_updates_available|Generic Sensor|
|nextcloud_system_apps_app_updates_calendar|Generic Sensor|
|nextcloud_system_apps_app_updates_contacts|Generic Sensor|
|nextcloud_system_apps_app_updates_tasks|Generic Sensor|
|nextcloud_system_storage_num_users|Generic Sensor|
|nextcloud_system_storage_num_files|Generic Sensor|
|nextcloud_system_storage_num_storages|Generic Sensor|
|nextcloud_system_storage_num_storages_local|Generic Sensor|
|nextcloud_system_storage_num_storage_home|Generic Sensor|
|nextcloud_system_storage_num_storages_other|Generic Sensor|
|nextcloud_system_shares_num_shares|Generic Sensor|
|nextcloud_system_shares_num_shares_user|Generic Sensor|
|nextcloud_system_shares_num_shares_groups|Generic Sensor|
|nextcloud_system_shares_num_shares_link|Generic Sensor|
|nextcloud_system_shares_num_shares_mail|Generic Sensor|
|nextcloud_system_shares_num_shares_room|Generic Sensor|
|nextcloud_system_shares_num_shares_link_no_password|Generic Sensor|
|nextcloud_system_shares_num_fed_shares_sent|Generic Sensor|
|nextcloud_system_shares_num_fed_shares_received|Generic Sensor|
|nextcloud_system_shares_permissions_3_1|Generic Sensor|
|nextcloud_server_webserver|Generic Sensor|
|nextcloud_server_php_version|Generic Sensor|
|nextcloud_server_php_memory_limit|Generic Sensor|
|nextcloud_server_php_max_execution_time|Generic Sensor (Maybe Timestamp Sensor?)|
|nextcloud_server_php_upload_max_filesize|Generic Sensor|
|nextcloud_server_database_type|Generic Sensor|
|nextcloud_server_database_version|Generic Sensor|
|nextcloud_server_database_version|Generic Sensor|
|nextcloud_activeusers_last5minutes|Generic Sensor|
|nextcloud_activeusers_last1hour|Generic Sensor|
|nextcloud_activeusers_last24hours|Generic Sensor|

Reply:
I've added a commit (47c5603)with a sample implementation of the Binary Sensors. I'm not sure if it's the best way to do it, but I've added a global variable with the names of the binary sensors. If there's a better way, please let me know.

Reply:
Good.